pkg/deploy/registry/deployconfig/strategy.go

```diff
@@ -39,14 +41,53 @@ func (strategy) AllowUnconditionalUpdate() bool {
 
 // PrepareForCreate clears fields that are not allowed to be set by end users on creation.
 func (strategy) PrepareForCreate(obj runtime.Object) {
-	_ = obj.(*api.DeploymentConfig)
-	// TODO: need to ensure status.latestVersion is not set out of order
+	dc := obj.(*api.DeploymentConfig)
+	dc.Generation = 1
+	dc.Status = api.DeploymentConfigStatus{}
 }
 
 // PrepareForUpdate clears fields that are not allowed to be set by end users on update.
 func (strategy) PrepareForUpdate(obj, old runtime.Object) {
-	_ = obj.(*api.DeploymentConfig)
-	// TODO: need to ensure status.latestVersion is not set out of order
+	newDc := obj.(*api.DeploymentConfig)
+	oldDc := old.(*api.DeploymentConfig)
+
+	newVersion := newDc.Status.LatestVersion
+	oldVersion := oldDc.Status.LatestVersion
+
+	// Check for imagechange controller change
+	isImageControllerChange := deployutil.IsImageChangeControllerChange(*newDc, *oldDc)
+
+	// Persist status
+	newStatus := newDc.Status
+	newDc.Status = oldDc.Status
+
+	// If this update comes from the imagechange controller, tolerate status detail updates.
+	// TODO: Make the config change controller detect this change and update latestVersion.
+	// Then the main controller should detect there is a change coming from the image change
+	// controller and update status details accordingly.
+	if isImageControllerChange {
+		newDc.Status.Details = newStatus.Details
+	}
+
+	// oc deploy --latest from new clients
+	if newVersion == oldVersion && deployutil.IsInstantiated(newDc) {
+		// TODO: Have an endpoint for this and avoid hacking it here.
```

Comment:
Ok - so no further concerns beyond those from me.  Do you have good soak on
this?

On Mon, May 9, 2016 at 11:28 AM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/deploy/registry/deployconfig/strategy.go
> <https://github.com/openshift/origin/pull/7149#discussion_r62518878>:
>
> > +
> > +	// Persist status
> > +	newStatus := newDc.Status
> > +	newDc.Status = oldDc.Status
> > +
> > +	// If this update comes from the imagechange controller, tolerate status detail updates.
> > +	// TODO: Make the config change controller detect this change and update latestVersion.
> > +	// Then the main controller should detect there is a change coming from the image change
> > +	// controller and update status details accordingly.
> > +	if isImageControllerChange {
> > +		newDc.Status.Details = newStatus.Details
> > +	}
> > +
> > +	// oc deploy --latest from new clients
> > +	if newVersion == oldVersion && deployutil.IsInstantiated(newDc) {
> > +		// TODO: Have an endpoint for this and avoid hacking it here.
>
> #6980 <https://github.com/openshift/origin/issues/6980>
> … <#m_3903626190098980304_>
> On Mon, May 9, 2016 at 5:19 PM, Clayton Coleman ***@***.***> wrote: In
> pkg/deploy/registry/deployconfig/strategy.go <#7149 (comment)
> <https://github.com/openshift/origin/pull/7149#discussion_r62517297>>: >
> + > + // Persist status > + newStatus := newDc.Status > + newDc.Status =
> oldDc.Status > + > + // If this update comes from the imagechange
> controller, tolerate status detail updates. > + // TODO: Make the config
> change controller detect this change and update latestVersion. > + // Then
> the main controller should detect there is a change coming from the image
> change > + // controller and update status details accordingly. > + if
> isImageControllerChange { > + newDc.Status.Details = newStatus.Details > +
> } > + > + // oc deploy --latest from new clients > + if newVersion ==
> oldVersion && deployutil.IsInstantiated(newDc) { > + // TODO: Have an
> endpoint for this and avoid hacking it here. Link an issue to the PR — You
> are receiving this because you were mentioned. Reply to this email directly
> or view it on GitHub <
> https://github.com/openshift/origin/pull/7149/files/dda90ccb39a505d676deab0f2c2864e6d8229ae3#r62517297>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7149/files/dda90ccb39a505d676deab0f2c2864e6d8229ae3#r62518878>
>